### PR TITLE
Return try again when full_history_ts_low is higher than requested ts

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR helps handle the race condition mentioned in this comment thread: https://github.com/facebook/rocksdb/pull/7884#discussion_r572402281 In case where actual full_history_ts_low is higher than the user's requested ts, return a try again message so they don't have the misconception that data between [ts, full_history_ts_low) is kept.

Test plan
```
$COMPILE_WITH_ASAN=1 make -j24 all
$./db_with_timestamp_basic_test --gtest_filter=UpdateFullHistoryTsLowTest.ConcurrentUpdate
$ make -j24 check
```

### CODE REVIEW

File: db/db_impl/db_impl_compaction_flush.cc [hunk 1]
```diff
@@ -950,21 +950,31 @@ Status DBImpl::IncreaseFullHistoryTsLow(ColumnFamilyHandle* column_family,
 Status DBImpl::IncreaseFullHistoryTsLowImpl(ColumnFamilyData* cfd,
                                             std::string ts_low) {
   VersionEdit edit;
+  std::string current_ts_low;
   edit.SetColumnFamily(cfd->GetID());
   edit.SetFullHistoryTsLow(ts_low);
 
   InstrumentedMutexLock l(&mutex_);
-  std::string current_ts_low = cfd->GetFullHistoryTsLow();
+  current_ts_low = cfd->GetFullHistoryTsLow();
```
Comment (riversand963) [inline db/db_impl/db_impl_compaction_flush.cc:958]: Why separate declaration and definition?

File: db/db_impl/db_impl_compaction_flush.cc [hunk 2]
```diff
@@ -950,21 +950,31 @@ Status DBImpl::IncreaseFullHistoryTsLow(ColumnFamilyHandle* column_family,
...
-  return versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(), &edit,
-                                &mutex_);
+  Status s = versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(),
+                                    &edit, &mutex_);
+  if (!s.ok()) {
+    return s;
+  }
+  current_ts_low = cfd->GetFullHistoryTsLow();
+  if (ucmp->CompareTimestamp(current_ts_low, ts_low) > 0) {
+    return Status::TryAgain(
```
Comment (riversand963) [inline db/db_impl/db_impl_compaction_flush.cc:973]: Nit: better print both timestamps (in hex) in error message to assist debugging.

File: db/db_with_timestamp_basic_test.cc [hunk 3]
```diff
@@ -366,6 +366,60 @@ TEST_F(DBBasicTestWithTimestamp, UpdateFullHistoryTsLowWithPublicAPI) {
...
+  SyncPoint::GetInstance()->LoadDependency({
+      {"DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "write_higher:0",
+       "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "WriteLowerStart"},
+      {"DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "WriteLowerStart",
+       "DBBasicTestWithTimestamp::UpdateFullHistoryTsLowWithPublicAPI:"
+       "write_higher:1"},
+  });
```
Comment (riversand963) [inline db/db_with_timestamp_basic_test.cc:402]: I think it's trickier trying to set up test sync points dependencies between two threads that execute exactly the same set of functions. In this case, I wonder if we can use the following hack to get rid of `LoadDependency()` here.
1. In DBImpl::IncreaseFullHistoryTsLowImpl(), set up a sync point with callback and use the address of the local variable `edit` as argument. When sync points are hit by the two writer threads, we can establish a mapping from `std::this_thread::get_id()` to address of version edit.
2. In the callback registered with VersionSet::LogAndApply:BeforeWriterWaiting, we can figure out which thread hits this point first. For the first comer, we can access the `VersionEdit` via address and set its `full_history_ts_low` to be 25. For the thread that comes next, we set `full_history_ts_low` to 10.
3. For both threads, assert the results of execution.

**Review by riversand963 — COMMENTED** [review]
Thanks @jowlyzhang for the PR. Left some comments on the unit test.

**Review by riversand963 — APPROVED** [review]
Thanks @jowlyzhang for the PR.
Please add a comment to `DB::IncreaseFullHistoryTsLow()` in include/rocksdb/db.h to mention when this can happen.

File: db/db_impl/db_impl_compaction_flush.cc (hunk 1 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_compaction_flush.cc]: Thank you for catching this. It's from a previous version of change, I forgot to put it back. It's reverted now.

File: db/db_with_timestamp_basic_test.cc [hunk 4]
```diff
@@ -366,6 +366,60 @@ TEST_F(DBBasicTestWithTimestamp, UpdateFullHistoryTsLowWithPublicAPI) {
   s = db_->IncreaseFullHistoryTsLow(db_->DefaultColumnFamily(),
                                     ts_low_str_null);
   ASSERT_EQ(s, Status::InvalidArgument());
+  // test IncreaseFullHistoryTsLow concurrently. When actual ts_low is higher
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/db_with_timestamp_basic_test.cc]: Thank you for the suggestion. This test is grouped into a separate test fixture now.

File: db/db_impl/db_impl_compaction_flush.cc (hunk 2 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_compaction_flush.cc]: Done

File: db/db_with_timestamp_basic_test.cc (hunk 3 above)
Comment (jowlyzhang) [reply db/db_with_timestamp_basic_test.cc]: This is a good idea. Thank you for helping check this and proposing this change.

I tried this workaround out. It worked well to eliminate some edge cases, but there are still more work using test sync points to coordinate the two writing threads to guarantee a TryAgain error. We discussed offline and decided to simplify the test logic with a different workaround.

File: include/rocksdb/db.h [hunk 5]
```diff
@@ -1421,6 +1421,9 @@ class DB {
 
   // Increase the full_history_ts of column family. The new ts_low value should
   // be newer than current full_history_ts value.
+  // If another thread updates full_history_ts_low concurrently to a higher
+  // timestamp than the requested ts_low, a try again error will be returned.
+  // Data between [ts_low, full_history_ts_low) is not kept.
```
Comment (riversand963) [inline include/rocksdb/db.h:1426]: Note that `IncreaseFullHistoryTsLow()` does not actually remove data, so I think we can remove the last sentence.

Comment (jowlyzhang) [reply include/rocksdb/db.h]: Cool, removed this line.

File: db/db_impl/db_impl_compaction_flush.cc [hunk 6]
```diff
@@ -952,19 +952,33 @@ Status DBImpl::IncreaseFullHistoryTsLowImpl(ColumnFamilyData* cfd,
...
 
-  return versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(), &edit,
-                                &mutex_);
+  Status s = versions_->LogAndApply(cfd, *cfd->GetLatestMutableCFOptions(),
+                                    &edit, &mutex_);
+  if (!s.ok()) {
+    return s;
+  }
+  current_ts_low = cfd->GetFullHistoryTsLow();
+  if (ucmp->CompareTimestamp(current_ts_low, ts_low) > 0) {
```
Comment (riversand963) [inline db/db_impl/db_impl_compaction_flush.cc:973]: Should we check `current_ts_low` is non-empty?

**Review by riversand963 — APPROVED** [review]

File: db/db_impl/db_impl_compaction_flush.cc (hunk 6 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_compaction_flush.cc]: Good idea, I added this check.
